x/farming/types/proposal.go

```diff
@@ -180,28 +197,48 @@ func (p *UpdateRequestProposal) Validate() error {
 	}
-	if _, err := sdk.AccAddressFromBech32(p.FarmingPoolAddress); err != nil {
-		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid farming pool address %q: %v", p.FarmingPoolAddress, err)
-	}
-	if _, err := sdk.AccAddressFromBech32(p.TerminationAddress); err != nil {
-		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid termination address %q: %v", p.TerminationAddress, err)
-	}
-	if p.StakingCoinWeights.Empty() {
-		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "staking coin weights must not be empty")
+	if p.FarmingPoolAddress != "" {
+		if _, err := sdk.AccAddressFromBech32(p.FarmingPoolAddress); err != nil {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid farming pool address %q: %v", p.FarmingPoolAddress, err)
+		}
 	}
-	if err := p.StakingCoinWeights.Validate(); err != nil {
-		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid staking coin weights: %v", err)
+	if p.TerminationAddress != "" {
+		if _, err := sdk.AccAddressFromBech32(p.TerminationAddress); err != nil {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid termination address %q: %v", p.TerminationAddress, err)
+		}
 	}
-	if ok := ValidateStakingCoinTotalWeights(p.StakingCoinWeights); !ok {
-		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "total weight must be 1")
+	if p.StakingCoinWeights != nil {
+		if p.StakingCoinWeights.Empty() {
+			return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "staking coin weights must not be empty")
+		}
+		if err := p.StakingCoinWeights.Validate(); err != nil {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid staking coin weights: %v", err)
+		}
+		if !ValidateStakingCoinTotalWeights(p.StakingCoinWeights) {
+			return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "total weight must be 1")
+		}
 	}
 	if p.StartTime != nil && p.EndTime != nil {
 		if !p.EndTime.After(*p.StartTime) {
 			return sdkerrors.Wrapf(ErrInvalidPlanEndTime, "end time %s must be greater than start time %s", p.EndTime, p.StartTime)
 		}
 	}
-	if p.IsForFixedAmountPlan() == p.IsForRatioPlan() {
-		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "only one of epoch amount or epoch ratio must be provided")
+	isForFixedAmountPlan := p.IsForFixedAmountPlan()
+	isForRatioPlan := p.IsForRatioPlan()
+	switch {
+	case isForFixedAmountPlan && isForRatioPlan:
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "at most one of epoch amount or epoch ratio must be provided")
+	case isForFixedAmountPlan:
+		if err := p.EpochAmount.Validate(); err != nil {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid epoch amount: %v", err)
+		}
+	case isForRatioPlan:
+		if !p.EpochRatio.IsPositive() {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "epoch ratio must be positive: %s", p.EpochRatio)
+		}
+		if p.EpochRatio.GT(sdk.OneDec()) {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "epoch ratio must be less than 1: %s", p.EpochRatio)
+		}
```

Comment:
This validation logic is duplicated on several validation functions. so we can refactor this and other similar cases
If you don't mind, I'll add a refactoring commit to this PR. @hallazzang @kogisin

Reply:
@dongsam Yes we can create a separate validation function and use it in those places. Sounds good to me.

Reply:
Sounds good to me as well.